content/blog/2022-12-01-epidata-v4.Rmd

```diff
@@ -0,0 +1,128 @@
+
+Additionally, the reduced size of the foreign keys compared to their text equivalents made it possible to include more [indexes](https://en.wikipedia.org/wiki/Database_index) in our database that let it quickly find data based on different subsets of criteria.  No matter how our users specify their requests, we have ways to look it up fast.
+
+![Fig 2: simplified diagram of v4 design, showing dimension tables with foreign key relationships and latest/full table scheme. [Full v4 schema details available on GitHub.](https://github.com/cmu-delphi/delphi-epidata/blob/v0.4.0/src/ddl/v4_schema.sql)](/blog/2022-12-01-epidata-v4_files/v4-table-design.png)
+
+## Performance results
+All performance figures were collected in an isolated QA environment. The core dataset was taken from a September, 2021 snapshot of the production Epidata database and used as-is for v3 experiments and converted to v4 format for v4 experiments. Performance experiments for v3 and v4 were both run on the same machine during dedicated time when nothing else was running. Input data for the "Adding new data" experiment was collected from data actually added to the production system in late May 2022. Input data for the query time experiments was sampled from a segment of actual query logs from July and August 2021, stratified by a generalized parameterized query template. Stratification ensured the samples were reasonably representative of actual traffic in terms of how many reference dates, regions, and signals are requested in a single query, as well as whether queries require the full data revision history or only the most recent issue of each reference date. Additional details about query log sampling will be included in a future post.
+
+Overall, the v4 system improves Epidata performance for adding new data and substantially improves performance for querying the most up-to-date figures. Storage performance is the same, although we suspect we could improve this by removing an index or two without much impact on query performance. Query performance for queries that must touch the data revision history is worse, but not by much; performance should still be adequate for most research use cases (*please do get in touch if this is not the case for you*).
+
+Metric | v4 performance relative to v3 |
+--- | --- | --- |
+Disk footprint | same |
+Adding new data | 30% faster |
+Querying latest data, i.e., the most recent issue of any reference date | 96% faster |
+Querying historical data, e.g. for as-of, issue, or lag queries | 120% slower |
+
+Query time (seconds) | Latest | Historical |
+--- | --- | --- |
+Percentile | v3 | v4 | v3 | v4 |
+0th (minimum) | 0.005 | 0.005 | 0.005 | 0 .005 |
+25th | 0.007 | 0.006 | 0.006 | 0.007 |
+50th (median) | 0.009 | 0.008 | 0.007 | 0.008 |
+75th | 0.058 | 0.017 | 0.012 | 0.023 |
+100th (maximum) | 739.332 | 20.069 | 17.121 | 15.997 |
```

Comment:
@minhkhul these tables are not super legible in the rendered website:
![image](https://user-images.githubusercontent.com/1158666/214077936-be3a541e-69c1-4f3b-9dab-0ce100a2ee60.png)

Can you experiment with styling options to resolve this? Goals are:
* easily tell which text belongs to which row (use borders, background color, whatever you think is good)
* reduce unnecessary line-wrapping

Reply:
<img width="1039" alt="Screenshot 2023-01-24 at 11 52 56 AM" src="https://user-images.githubusercontent.com/118945681/214356581-b9f46940-0d8a-4fb5-94f7-86a80e0adc9b.png">